content/tutorials/parallelization/SIMWE_parallelization.qmd

```diff
@@ -0,0 +1,804 @@
+            tools.r_mapcalc(expression=f"detachment_coef_{huc12} = 0.001")
+            tools.r_mapcalc(expression=f"shear_stress_{huc12} = 0.01")
+            region = tools.g_region(flags="p", format="json")
+            tools.r_sim_sediment(
+                elevation="ned",
+                water_depth=f"depth_{huc12}",
+                dx=f"dx_{huc12}",
+                dy=f"dy_{huc12}",
+                detachment_coeff=f"detachment_coef_{huc12}",
+                transport_coeff=f"transport_capacity_{huc12}",
+                shear_stress=f"shear_stress_{huc12}",
+                niterations=simulation_time,
+                erosion_deposition=f"erdep_tmp_{huc12}",
+                nwalkers=5 * region["cells"],
+            )
+            tools.r_grow(
+                input=f"erdep_tmp_{huc12}",
+                output=f"erdep_{huc12}",
+                radius=-2.01,
+            )
+            # Erosion: extract negative values and convert to positive mass [kg/m^2]
+            tools.r_mapcalc(
+                expression=f"erosion_{huc12} = if(erdep_{huc12} < 0, abs(erdep_{huc12}) * {simulation_time * 60}, 0)",
+            )
+            erosion = tools.r_univar(map=f"erosion_{huc12}", format="json", flags="e")
+            return {
+                "huc12": huc12,
+                "erosion_mean": erosion["mean"],
+                "erosion_median": erosion["median"],
+                "erosion_total": erosion["sum"] * region["nsres"] * region["ewres"],
+            }
+
+
+if __name__ == "__main__":
+    tools = Tools()
+    # The entire workflow will run in parallel,
+    # so this limits the number of threads the tools can use to 1.
+    tools.g_gisenv(set="NPROCS=1")
+    basins = tools.v_db_select(format="json", map="river_basins")["records"]
+    huc12s = [basin["huc12"] for basin in basins]
+    # set the number of processes:
+    with Pool(processes=cpu_count()) as pool:
+        result = list(tqdm(pool.imap(compute, huc12s), total=len(huc12s)))
+        with open("result.json", "w") as fp:
+            json.dump(result, fp)
+```
+
+Now execute the script, it will take some time.
+
+```{python}
+%run script.py
+```
+
+Load the resulting JSON file into a pandas dataframe and normalize the results, so that we can easily compare the subwatersheds.
+
+```{python}
+import pandas as pd
+import json
+
+with open("result.json") as f:
+    stats = json.load(f)
+
+df = pd.DataFrame(stats)
+df["normalized_erosion"] = df.erosion_mean / max(df.erosion_mean)
+df
+```
+
+Load the subwatershed layers into geopandas for visualization. Join the dataframe with the simulation values using the _huc12_ key and keep only subwatersheds that have computed erosion values.
+
+```{python}
+import geopandas as gpd
+
+gdf = gpd.read_file(hydro_filename, layer="WBDHU12")
+gdf = gdf.merge(df, on="huc12", how="left")
+# JSON does not support Timestamp type
+gdf = gdf.drop(columns=["loaddate"])
+gdf = gdf[gdf["erosion_total"].notna()]
```

Comment:
Just comment what this is doing. Remove results without erosion? But why are these here anyway?